python/ql/src/experimental/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -618,6 +709,22 @@ predicate comprehensionReadStep(CfgNode nodeFrom, Content c, EssaNode nodeTo) {
   )
 }
 
+/**
+ * In
+ * ```python
+ * obj.foo
+ * ```
+ * data flows from `obj` to `obj.foo` via a read from `foo`.
+ */
+predicate attributeReadStep(CfgNode nodeFrom, Content c, CfgNode nodeTo) {
+  exists(AttrNode attr |
+    nodeTo.asCfgNode() = attr and
+    nodeFrom.asCfgNode() = attr.getObject() and
+    attr.getName() = c.(AttributeContent).getAttribute() and
+    attr.isLoad()
+  )
+}
```

Comment:
```suggestion
/**
 * Holds if `nodeTo` is a read of an attribute (corresponding to `c`) of the object in `nodeFrom`.
 *
 * For example, in
 * ```python
 * obj.foo
 * ```
 * data flows from `obj` to `obj.foo` via a read from `foo`.
 */
predicate attributeReadStep(CfgNode nodeFrom, AttributeContent c, CfgNode nodeTo) {
  exists(AttrNode attr |
    nodeFrom.asCfgNode() = attr.getObject() and
    nodeTo.asCfgNode() = attr and
    attr.getName() = c.getAttribute() and
    attr.isLoad()
  )
}
```
A few minor rewrite suggestions. It's more readable to me to see `nodeFrom` before `nodeTo`. Also, if we only ever cast `c` to `AttributeContent`, we might as well restrict the type of `c` in advance.